### PR TITLE
Add test for folder generation

### PR DESCRIPTION
**Pull request by killianrochet**

### CODE REVIEW

File: bep032tools/generator/BEP032Generator.py [hunk 1]
```diff
@@ -267,7 +267,7 @@ def validate(self):
         bep032tools.validator.BEP032Validator.is_valid(self.basedir)
 
     @classmethod
-    def generate_struct(cls, csv_file, pathToDir):
+    def generate_struct(cls, csv_file, pathToDir):
```
Comment (JuliaSprenger) [inline bep032tools/generator/BEP032Generator.py:270]: I think this change is unintentional, right? Could you reverse it?

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import os
+import pathlib
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:2]: You are not using this import in your test. Maybe you forgot this in your cleanup?

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 3]
```diff
@@ -184,7 +185,32 @@ def test_generate_example_structure(self):
...
+            exist_path = True
+        print(f" folder created ? {create_folder}")
+        self.assertTrue(exist_path)
 
 if __name__ == '__main__':
     unittest.main()
 
+
+
+
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:216]:
```suggestion
```

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 4]
```diff
@@ -184,7 +185,42 @@ def test_generate_example_structure(self):
...
+    def setUp(self):
+        self.test_dir = Path(initialize_test_directory(clean=True))
+        os.mkdir(self.test_dir / 'generateTest')
+        csv_filename = generate_simple_csv_file()
+        self.csv_file = csv_filename
+
+    def test_generate_folder(self):
+        """
+        Test if the Path exists. After generation, check that the folder exists and display the list of sub-folders.
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:198]: Hi @killianrochet, i think the test your are aiming to implement should be more specific. You can read again the issue you raised here (#123) and try to cover the two cases your were describing in that issue (potentially in 2 separate `test_` methods.
Within such a test method you should 1) try to generate the situation in which the error was observed 2) run the generator and 3) assert that the observed error does not occur any more.

So a practical next step would be to split this test in 2 methods `test_no_duplicate_folder_generated` and `test_generate_folder` (to ensure the error you reported  if the folder does not yet exist does not occur).

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 5]
```diff
@@ -184,7 +185,42 @@ def test_generate_example_structure(self):
...
+        Parameters
+        ----------
+        Returns
+        ----------
+        bool
+            True if validation was successful. False if it failed.
+        """
+        self.test_dir = self.test_dir / 'test'
+        generation = False
+        if Path(self.test_dir).exists():
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:208]: The testsuite executes the `setUp` before each run of a `test_` method. In the `setUp` we are initializing the `self.test_dir` (via `initialize_test_directory`). So checking here again if this folder exists is duplicate and the `else` case of your condition below will never be executed.

Comment (killianrochet) [reply bep032tools/generator/tests/test_BEP032Generator.py:208]: Okay, I'll change that right now

File: bep032tools/generator/tests/test_BEP032Generator.py (hunk 4 above)
Comment (killianrochet) [reply bep032tools/generator/tests/test_BEP032Generator.py:198]: Ok, I have separated the different tests into two different test functions

**Comment by JuliaSprenger** [comment]
The next step is that in each of your test functions you create the situation in which the bug occurred. For this it would help to put a description of this in the docstring. Roughly like `Ensure that in situation A behavior B does not happen`.

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 6]
```diff
@@ -184,7 +184,56 @@ def test_generate_example_structure(self):
...
+        ----------
+        Returns
+        ----------
+        bool
+            True if there are no duplicates. False if a duplicate exists
+        """
+        generation = False
+        if os.path.isdir(self.test_dir) and os.path.dirname(self.test_dir) not in os.listdir(self.test_dir):
+            generation = True
+        self.assertTrue(generation)
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:231]: Yes, this is getting closer to what the test should look like. You have code to check for the existence of folders, the only thing that is missing is that you actually run the method that you want to test in your test. You need to call `BEP032Data.generate_struct(self.csv_file, self.test_dir)` somewhere before checking that no duplicate subfolders are generated.

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 7]
```diff
@@ -184,7 +184,56 @@ def test_generate_example_structure(self):
...
+    def setUp(self):
+        self.test_dir = Path(initialize_test_directory(clean=True))
+        os.mkdir(self.test_dir / 'generateTest')
+        csv_filename = generate_simple_csv_file()
+        self.csv_file = csv_filename
+
+    def test_generate_folder(self):
+        """
+        When the file did not exist, the generation did not work properly.
+        Now if the folder does not exist, the function creates the folder before the generation function is launched
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:198]: This is method here is a test and will not be run when a user is using the BEP032Generator. In your documentation it sounds like you expect this code here to be executed also in the normal user mode. This should be rather something like `Check that generate_struct does also accepts non-existing target folders and generates missing folders correctly.`

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 8]
```diff
@@ -184,7 +184,56 @@ def test_generate_example_structure(self):
...
+        ----------
+        bool
+            True if the generation is well done. False if it failed.
+        """
+        self.test_dir = self.test_dir / 'test'
+        test_generate = False
+        if os.path.isdir(self.test_dir):
+            BEP032Data.generate_struct(self.csv_file, self.test_dir)
+            test_generate = True
+        else:
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:211]: this else condition will never be run, as the `self.test_dir` folder is always generated by the `setUp` method before running the test.

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 9]
```diff
@@ -184,7 +184,36 @@ def test_generate_example_structure(self):
...
+    def test_generate_folder(self):
+        """
+        When the folder did not exist, the generation did not work properly.
+        Now if the folder does not exist, the function creates the folder before the generation function is launched
+        """
+        test_generate = False
+        BEP032Data.generate_struct(self.csv_file, self.test_dir)
+        if os.path.isdir(self.test_dir):
+            test_generate = True
+        self.assertTrue(test_generate)
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:204]: Yes, now the test is doing what it's supposed to do :+1:

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 10]
```diff
@@ -184,7 +184,36 @@ def test_generate_example_structure(self):
     def doCleanups(self):
         initialize_test_directory(clean=True)
 
+class Test_FolderGeneration(unittest.TestCase):
+
+    def setUp(self):
+        self.test_dir = Path(initialize_test_directory(clean=True))
+        self.test_dir = self.test_dir / 'generateTest'
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:191]: It's bad practice to overwrite variables immediately after setting them. I suggest also a renaming here
```suggestion
        test_dir = Path(initialize_test_directory(clean=True))
        self.test_dir = test_dir / 'generateTest'  # this folder will not exist yet
```

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 11]
```diff
@@ -184,7 +184,36 @@ def test_generate_example_structure(self):
...
+        self.assertTrue(test_generate)
+
+    def test_no_duplicate_folder_generated(self):
+        """
+        Checks that there are no duplicates when generating folders.
+        If a duplicate exists and has the same name as the parent folder, the function returns an error
+        """
+        BEP032Data.generate_struct(self.csv_file, self.test_dir)
+        generation = False
+        if os.path.isdir(self.test_dir) and os.path.basename(self.test_dir) not in os.listdir(self.test_dir):
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:213]: The check for the generation of the root folder we already covered in the other test function, so we do not need to reimplement that here. Also you know exactly the filename you want to check for, so you don't need to iterate all existing folders.
```suggestion
    root_name = self.test_dir.name
        if  not (self.test_dir / root_name).exists()):
```

**Comment by JuliaSprenger** [comment]
@killianrochet Last thing to fix: PEP8 (see pep8speaks comment above)

File: bep032tools/generator/tests/test_BEP032Generator.py [hunk 12]
```diff
@@ -185,6 +185,35 @@ def doCleanups(self):
...
+        generation = False
+        root_name = self.test_dir.name
+        if not (self.test_dir / root_name).exists():
+            generation = True
+        self.assertTrue(generation)
+
+
 if __name__ == '__main__':
     unittest.main()
-
```
Comment (JuliaSprenger) [inline bep032tools/generator/tests/test_BEP032Generator.py:190]: @killianrochet This change is also not intended, right?

Comment (JuliaSprenger) [reply bep032tools/generator/tests/test_BEP032Generator.py:190]: Or was this required by Pep8?

**Review by JuliaSprenger — APPROVED** [review]

**Comment by JuliaSprenger** [comment]
@killianrochet And you can mark the PR as ready for review. It's about time ;)

File: bep032tools/generator/tests/test_BEP032Generator.py (hunk 12 above)
Comment (killianrochet) [reply bep032tools/generator/tests/test_BEP032Generator.py:190]: It was required by Pep8 and I don't know why to be honest

Comment (JuliaSprenger) [reply bep032tools/generator/tests/test_BEP032Generator.py:190]: ok, then we should keep it

**Comment by JuliaSprenger** [comment]
Cool :) Thanks for completing this! I will merge it now. :champagne:

**Pull request merged**